indigo_app/static/javascript/indigo/grammars.js

```diff
@@ -0,0 +1,200 @@
+        .firstChild.textContent
+        // remove multiple consecutive blank lines
+        .replace(/^( *\n){2,}/gm, "\n");
+    }
+
+    /**
+     * Configure a new instance of a Monaco editor.
+     */
+    setupEditor (editor) {
+      this.installActions(editor);
+    }
+
+    installLanguage () {
+      monaco.languages.register({ id: this.language_id });
+      monaco.languages.setMonarchTokensProvider(this.language_id, this.language_def);
+      monaco.editor.defineTheme(this.theme_id, this.theme_def);
+    }
+
+    /**
+     * Install named actions on the editor. Subclasses should install at least these actions:
+     *
+     * * format.bold
+     * * format.italics
+     * * format.remork
+     * * insert.schedule
+     * * insert.table
+     *
+     * @param editor monaco editor instance
+     */
+    installActions (editor) {
+    }
+
+    /**
+     * Markup a textual remark
+     */
+    markupRemark (text) {
+    }
+
+    /**
+     * Markup a link (ref)
+     */
+    markupRef (title, href) {
+    }
+
+    /**
+     * Markup an image
+     */
+    markupImage (title, src) {
+    }
+
+    insertRemark (editor, remark) {
+      const sel = editor.getSelection();
+      editor.executeEdits(this.language_id, [{
+        identifier: 'insert.remark',
+        range: sel,
+        text: this.markupRemark(remark),
+      }]);
+      editor.pushUndoStop();
+    }
+
+    /**
+     * Get a description of the image at the cursor, if any.
+     */
+    getImageAtCursor (editor) {
+      const match = this.getMatchAtCursor(editor, this.image_re);
+      if (match) {
+        return {
+          match: match,
+        };
+      }
+    }
+
+    /**
+     * Insert or update the image at the current cursor
+     */
+    insertImageAtCursor (editor, filename) {
+      const existing = this.getImageAtCursor(editor);
+      let sel = editor.getSelection();
+      let image;
+
+      if (existing) {
+        // update existing image
+        image = this.markupImage(existing.title, filename);
+        sel = sel.setEndPosition(sel.startLineNumber, existing.match.index + existing.match[0].length + 1);
+        sel = sel.setStartPosition(sel.startLineNumber, existing.match.index + 1);
+      } else {
+        // insert new image
+        image = this.markupImage('', filename);
+      }
+
+      editor.executeEdits(this.language_id, [{
+        identifier: 'insert.image',
+        range: sel,
+        text: image,
+      }]);
+
+      editor.pushUndoStop();
+    }
+
+    /**
+     * Returns the match object, if any, for a regular expression match at the current cursor position,
+     * on the current line.
+     */
+    getMatchAtCursor (editor, regexp) {
+      const sel = editor.getSelection();
+      const line = editor.getModel().getLineContent(sel.startLineNumber);
+
+      for (let match of line.matchAll(regexp)) {
+        if (match.index <= sel.startColumn && sel.startColumn <= match.index + match[0].length) {
```

Comment:
I _think_ I can see what's happening here but might a comment be a good idea?

Reply:
Added